### DISCUSSION TITLE
Designing the API

### DISCUSSION THREAD

**Discussion by sbenthall** [scikit-agent/scikit-agent#81] [General]
What are the objects that we need to define as we build out the library?
What should the API be?

- **Blocks**. Block are collections of structural equations with exogenous shocks. They represent a Structural Causal Model, or Structural Causal Game.
  - Blocks can be composed with other blocks. Currently, DBlocks can be composed into an RBlock 
  - The structural equations are governed by **Rules**, which can be expressed in strings (text), sympy, lambdas. 
- "Bellman" **Objects** and **Functions**.  An object that carries that transition functions, decision functions, and reward functions. These are from (arrival) states, "to" post-states (which may or may not be t+1 versions of the arrival states).
  - **Trainable** versions of these "Bellman" functions: BlockPolicyNet, BlockValueNet..... Given a loss function, tune the parameters of these functions to minimize loss.
- **Loss functions**. These are for training neural nets and other optimizers. The include: expected lifetime reward ($T \in [1, \infty)$); Bellman residual loss function (which gets propagated back to both the Policy and Value functions)...
- **Solution Algorithms**. These combine more basic elements into a verified strategy for computing optimal and equilibrium decision rules.
- **Grids**. Labeled arrays or tables of data which are used in training, for approximation, etc..

What else? Simulation? Emergence? Estimation?

Maybe:
- (We would like) **Models**: compositions of blocks in nontrivials. E.g. Block1 once, then 5 of Block2, then Block 3. Or, Block1 with stochastic transition to Block2.  Or, Block1, then infinite Block2.

**Comment by sbenthall** [comment]
@alanlujan91 looping back to this "Discussion" thread after thinking about some of what we discussed in our meeting yesterday, and some prior meetings.

My understanding is that you would like, for a variety of good reasons, there to be something in *scikit-agent* that represents what might be called a _sequential Bellman problem_: a dynamic problem which is subdivided into "stages" that have some useful properties (a single agent per stage; a clear sequence of continuation values; soluble via dynamic programming methods).

As I think we've noted, the design that's implemented at this time doesn't quite fit that those expectations. The Blocks are a more general way of representing models (multiple agents, potentially not sequentially divisible, potentially not soluble via dynamic programming).

But there are conditions under which Blocks can be easily converted into Bellman objects and I think many cases when we would like to do just that.

I'm wondering if there's another kind of object that can be put into this taxonomy that would accommodate you design vision here.
For example, we could imagine that a subset of Bellman objects are really Bellman 'stages': they don't require the arrival and departure states to be the same. They have only a single agent. And so on.

If that were well-defined, we could also implement something that converts Blocks to this form (and throws an error if it can't be converted to that form).

Does that sound right? What are the formal properties of the kind of object that you have in mind? Can you write them in math?
